### PR TITLE
Feature/author write

### PR DESCRIPTION
**Pull request by galiri**
this is a version that has enough functionality for J3O. 
TODO includes:
1.when auhtor service fails to load authors- stop write?

2. also in the request for author transformer we may need to separate the  environment  the call is made to, I put the full URL in as one programme  arg.

### CODE REVIEW

File: health/healthcheck.go [hunk 1]
```diff
@@ -35,37 +58,59 @@ func (service *HealthService) healthChecker() (string, error) {
...
 		return "Could not connect to elasticsearch", err
 	}
-
 	return "Successfully connected to the cluster", nil
 }
 
+func (service *HealthService) v1AuthorsTransformerConnectivityCheck() fthealth.Check {
+	return fthealth.Check{
+		ID:               "check-connectivity-to-v1-authors-transformer",
+		BusinessImpact:   "It is not possible to identify FT authors in People",
```
Comment (peteclark-ft) [inline health/healthcheck.go]: Maybe reword this to "Impacts our ability to identify FT Authors from People"

File: health/healthcheck_test.go [hunk 2]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
 	}
+
+	esHealthService.AssertExpectations(t)
+	authorService.AssertExpectations(t)
 }
 
-func TestHealthServiceConnectivityChecker(t *testing.T) {
+func TestGoodToGoUnhealthyV1AuthorsTransformer(t *testing.T) {
+	//create a request to pass to our handler
+	req, err := http.NewRequest("GET", "/__gtg", nil)
```
Comment (peteclark-ft) [inline health/healthcheck_test.go]: No need to change, but for future reference - `httptest.NewRequest` will not `err`.

File: health/healthcheck_test.go [hunk 3]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
+	if err != nil {
```
Comment (peteclark-ft) [inline health/healthcheck_test.go]: Similarly, you can do `require.NoError(t, err)` instead

File: health/healthcheck_test.go [hunk 4]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
-	assert.Equal(t, "Could not connect to elasticsearch", message)
-	assert.NotNil(t, err)
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(healthService.HealthCheckHandler())
+
+	handler.ServeHTTP(rr, req)
+	assert.Equal(t, http.StatusOK, rr.Code, "HealthCheck should return HTTP 200 OK")
 
+	checks, err := parseHealthcheck(rr.Body.String())
+	assert.NoError(t, err, "HealthCheck Response BOdy should be consistent")
```
Comment (peteclark-ft) [inline health/healthcheck_test.go]: Nitpicky but `BOdy` => `Body`

File: health/healthcheck_test.go [hunk 5]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
+	healthService := NewHealthService(esHealthService, authorService)
+
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(healthService.HealthCheckHandler())
+
+	handler.ServeHTTP(rr, req)
+	assert.Equal(t, http.StatusOK, rr.Code, "HealthCheck should return HTTP 200 OK")
+
+	checks, err := parseHealthcheck(rr.Body.String())
+	assert.NoError(t, err, "HealthCheck Response BOdy should be consistent")
```
Comment (peteclark-ft) [inline health/healthcheck_test.go]: `BOdy` again

File: health/healthcheck_test.go [hunk 6]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
```
Comment (peteclark-ft) [inline health/healthcheck_test.go]: And again `BOdy`

File: main.go [hunk 7]
```diff
@@ -90,6 +89,19 @@ func main() {
 		Desc:   "List which are currently supported by elasticsearch (already have mapping associated)",
 		EnvVar: "ELASTICSEARCH_WHITELISTED_CONCEPTS",
 	})
+	authorIdsURL := app.String(cli.StringOpt{
+		Name:   "authorIdsUrl",
```
Comment (peteclark-ft) [inline main.go]: All the other option names are in `-` syntax not camel case.

File: main.go [hunk 8]
```diff
@@ -90,6 +89,19 @@ func main() {
...
+		Value:  "http://localhost:8080/__v1-authors-transformer/",
+		Desc:   "The URL of authors ids endpoint  used to identify authors",
+		EnvVar: "AUTHOR_IDS_URL",
+	})
+
+	authorCredKey := app.String(cli.StringOpt{
+		Name:   "pub-cluster-cred",
```
Comment (peteclark-ft) [inline main.go]: `pub-cluster-credentials` will be clearer if anyone uses this option via the CLI

File: service/author_service.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	"fmt"
+	"net/http"
+	"strings"
+
+	transactionid "github.com/Financial-Times/transactionid-utils-go"
+	log "github.com/Sirupsen/logrus"
+)
+
+const contentType = "application/json"
+const idsPath = "/transformers/authors/__ids"
```
Comment (peteclark-ft) [inline service/author_service.go]: Since this is package level, maybe this should specifically be `authorTransformerIdsPath`?

File: service/author_service.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+const gtgPath = "/__gtg"
+
+type AuthorUUID struct {
```
Comment (peteclark-ft) [inline service/author_service.go]: Do we need this to be exported?

File: service/author_service_test.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+		assert.Equal(t, "UPP concept-rw-elasticsearch", ua, "user-agent header")
+		tid := req.Header.Get("X-Request-Id")
+		assert.NotEmpty(t, tid, "transactio id")
+		contentType := req.Header.Get("Content-Type")
+		user, password, _ := req.BasicAuth()
+
+		w.WriteHeader(m.Ids(contentType, user, password))
+
+		authorIds := `{"ID":"2916ded0-6d1f-4449-b54c-3805da729c1d"}
+{"ID":"ddc22d37-624a-4a3d-88e5-ba508e38c8ba"}`
```
Comment (peteclark-ft) [inline service/author_service_test.go]: (I worked this out late yesterday). This can be done on one line but not with backticks, you just need to escape all the `"` 

i.e.

```
{\"ID\":\"2916ded0-6d1f-4449-b54c-3805da729c1d\"}\n{\"ID\":\"ddc22d37-624a-4a3d-88e5-ba508e38c8ba\"}
```

File: service/es_model_populater.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
+package service
+
+import (
+	//	log "github.com/Sirupsen/logrus"
```
Comment (peteclark-ft) [inline service/es_model_populater.go]: This can be removed

File: health/healthcheck.go (hunk 1 above)
Comment (khatton-ft) [inline health/healthcheck.go]: Or possibly, "Updates that identify FT authors cannot be processed", as the impact is that no changes will be made. Known authors will still be boosted.

File: health/healthcheck_test.go [hunk 13]
```diff
@@ -140,43 +166,198 @@ func TestGoodToGoUnhealthyCluster(t *testing.T) {
...
+	// directly and pass in our Request and ResponseRecorder.
+	handler.ServeHTTP(rr, req)
 
+	// Series of verifications:
+	if status := rr.Code; status != http.StatusServiceUnavailable {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusServiceUnavailable)
+	}
+
+	if rr.Body.Bytes() != nil {
```
Comment (khatton-ft) [inline health/healthcheck_test.go]: "a negative GTG response (that with a response status other than 200) must provide a [text/plain; charset=US-ASCII] response body describing the nature of the problem" (from https://docs.google.com/document/d/11paOrAIl9eIOqUEERc9XMaaL3zouJDdkmb-gExYFnw0/edit), so this test is wrong.

File: service/author_service.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+	LoadAuthorIdentifiers() error
+	IsFTAuthor(UUID string) bool
+	IsGTG() error
+}
+
+//uses v1 transformer author list
+type curatedAuthorService struct {
+	httpClient             *http.Client
+	serviceURL             string
+	authorUUIDs            map[string]struct{}
```
Comment (khatton-ft) [inline service/author_service.go]: `struct{}` or `interface{}`?

File: service/author_service.go (hunk 10 above)
Comment (khatton-ft) [inline service/author_service.go]: I think if you want to unmarshal JSON into it, it may need to be exported.

File: service/author_service_test.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	"ddc22d37-624a-4a3d-88e5-ba508e38c8ba": struct{}{},
+}
+
+func (m *mockAuthorTransformerServer) startMockAuthorTransformerServer(t *testing.T) *httptest.Server {
+	r := mux.NewRouter()
+	r.HandleFunc(idsPath, func(w http.ResponseWriter, req *http.Request) {
+		ua := req.Header.Get("User-Agent")
+		assert.Equal(t, "UPP concept-rw-elasticsearch", ua, "user-agent header")
+		tid := req.Header.Get("X-Request-Id")
+		assert.NotEmpty(t, tid, "transactio id")
```
Comment (khatton-ft) [inline service/author_service_test.go]: :pencil2: `transactio id`

File: service/es_model_populater.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+	"fmt"
+
+	"github.com/Financial-Times/neo-model-utils-go/mapper"
+)
+
+const (
+	PERSON = "people"
+)
+
+type ModelPopulater interface {
```
Comment (khatton-ft) [inline service/es_model_populater.go]: I would spell this "Populator" but maybe that's just me.

File: service/es_model_populater.go [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	return reverse(conceptTypes)
+}
+
+func reverse(strings []string) []string {
+	if strings == nil {
+		return nil
+	}
+	if len(strings) == 0 {
+		return []string{}
```
Comment (khatton-ft) [inline service/es_model_populater.go]: You could just return `strings` here, strictly there's no need to create a new slice.

File: vendor/vendor.json [hunk 18]
```diff
@@ -47,27 +47,24 @@
...
 		},
 		{
-			"checksumSHA1": "ddLGmQTKn5BeEsKYFOGAH33sEfg=",
+			"checksumSHA1": "+sjiZBzrsk91W1A3Md1azCkZadg=",
 			"path": "github.com/Financial-Times/transactionid-utils-go",
-			"revision": "a2b1163e7c2f6c77add74ca2c7fc7422323a2d78",
-			"revisionTime": "2016-02-12T16:38:35Z",
-			"version": "0.1.0",
-			"versionExact": "0.1.0"
+			"revision": "df2f00c734957c9dd651ce23ab0e0902504c7636",
```
Comment (khatton-ft) [inline vendor/vendor.json]: Does this version not have a tag?

File: main.go [hunk 19]
```diff
@@ -119,13 +131,19 @@ func main() {
...
 
 		esService := service.NewEsService(ecc, *indexName, &bulkProcessorConfig)
-		var allowedConceptTypes []string = strings.Split(*elasticsearchWhitelistedConceptTypes, ",")
-		handler := resources.NewHandler(esService, allowedConceptTypes)
+
+		allowedConceptTypes := strings.Split(*elasticsearchWhitelistedConceptTypes, ",")
+		authorService, err := service.NewAuthorService(*authorIdsURL, *authorCredKey, &http.Client{Timeout: time.Second * 30})
+		if err != nil {
+			log.Errorf("Could not retrieve author list, error=[%s]\n", err)
+			//TODO we need to stop writing until we have authors
```
Comment (peteclark-ft) [inline main.go]: Should we stop writing? Or should we write it with `author: false` and then trigger a re-ingest once the issue is resolved?

File: mappings/mapping-concepts-0_0_2.json [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+              }]
+            }
+          }
+        },
+        "isFTAuthor": {
+          "type": "string",
+          "analyzer": "standard",
+          "fields": {
+            "raw": {
+              "type": "string",
```
Comment (peteclark-ft) [inline mappings/mapping-concepts-0_0_2.json]: Is there a type of boolean? Or will that not work with the category context above?

File: service/author_service.go (hunk 10 above)
Comment (peteclark-ft) [inline service/author_service.go]: I'm pretty sure only the fields within it need to be exported, but the type itself can be private?

File: mappings/mapping-concepts-0_0_2.json (hunk 20 above)
Comment (galiri) [inline mappings/mapping-concepts-0_0_2.json]: I There is a boolean  and it didn't work with the query when I tried it.

File: service/author_service.go (hunk 14 above)
Comment (galiri) [inline service/author_service.go]: apparently  a struct is more efficient. source= Dr Panziera

Comment (khatton-ft) [inline service/author_service.go]: `struct{}` is correct, I had mis-remembered.

File: main.go [hunk 21]
```diff
@@ -89,18 +89,18 @@ func main() {
 		Desc:   "List which are currently supported by elasticsearch (already have mapping associated)",
 		EnvVar: "ELASTICSEARCH_WHITELISTED_CONCEPTS",
 	})
-	authorIdsURL := app.String(cli.StringOpt{
-		Name:   "authorIdsUrl",
-		Value:  "http://localhost:8080/__v1-authors-transformer/",
+	pubClusterReadURL := app.String(cli.StringOpt{
+		Name:   "pub-cluster-read-url",
+		Value:  "http://localhost:8080",
 		Desc:   "The URL of authors ids endpoint  used to identify authors",
```
Comment (khatton-ft) [inline main.go]: The description of this parameter doesn't sound right now.

File: main.go [hunk 22]
```diff
@@ -90,14 +90,14 @@ func main() {
...
+		Name:   "publish-cluster-read-url",
 		Value:  "http://localhost:8080",
-		Desc:   "The URL of authors ids endpoint  used to identify authors",
+		Desc:   "The URL of the publish cluster (used for retreiving author list)",
 		EnvVar: "PUBLISH_CLUSTER_URL",
 	})
 
 	pubClusterCredKey := app.String(cli.StringOpt{
-		Name:   "pub-cluster-credentials",
+		Name:   "pubish-cluster-credentials",
```
Comment (khatton-ft) [inline main.go]: Typo
